### PR TITLE
Introduce attachments to Request API.

### PR DESCRIPTION
**Pull request by jkozlowski**
## Before this PR

There is not way to attach extra internal state that the channels can share and pass along the request.

## After this PR
==COMMIT_MSG==
Request.attachments() allows channels to attach additional state.
==COMMIT_MSG==

### Why is this mutable you ask?

Well, mutable is not-in vogue, but with the current design of channels, it might be difficult to associate any sort of mutable state per-request. For example, say when we retry, we wanted to make sure we first tried on hosts that we didn't try before, we couldn't really do that with mutable RequestAttachments, as channels get passed a request, and return a response: if I channel makes changes to the request, this will not be visible to upstream channels.

### CODE REVIEW

File: dialogue-target/src/main/java/com/palantir/dialogue/Request.java [hunk 1]
```diff
@@ -139,16 +148,21 @@ public static Builder builder() {
 
         private Optional<RequestBody> body = Optional.empty();
 
+        private RequestAttachments attachments;
+
         private int mutableCollectionsBitSet = 0;
 
-        private Builder() {}
+        private Builder() {
+            attachments = DefaultRequestAttachments.create();
```
Comment (carterkozak) [inline dialogue-target/src/main/java/com/palantir/dialogue/Request.java]: nit: let's keep this null and initialize in `build()`. That way `Builder.from(Request)` doesn't create objects to throw away.

File: dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachments.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+package com.palantir.dialogue;
+
+import javax.annotation.Nullable;
+
+public interface RequestAttachments {
+
+    <V> V put(RequestAttachmentKey<V> key, V value);
+
+    <V> V getOrDefault(RequestAttachmentKey<V> key, @Nullable V defaultValue);
```
Comment (carterkozak) [inline dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachments.java]: Should both methods be annotated `@Nullable`?

File: dialogue-target/src/main/java/com/palantir/dialogue/Request.java (hunk 1 above)
Comment (jkozlowski) [reply dialogue-target/src/main/java/com/palantir/dialogue/Request.java]: ah, misread the code, for some reason I took from as an overloaded constructor.

Comment (jkozlowski) [reply dialogue-target/src/main/java/com/palantir/dialogue/Request.java]: DOne

File: dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachments.java (hunk 2 above)
Comment (jkozlowski) [reply dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachments.java]: I don't think we want to support putting null, unless you want to reuse this api for remove?

Comment (carterkozak) [reply dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachments.java]: If there's no existing value, won't `put` return null?

Comment (jkozlowski) [reply dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachments.java]: oh I thought you wanted to annotate the value; yeah I should annotate the return, you're right.

Comment (jkozlowski) [reply dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachments.java]: Done

File: dialogue-target/src/main/java/com/palantir/dialogue/Request.java (hunk 1 above)
Comment (jkozlowski) [reply dialogue-target/src/main/java/com/palantir/dialogue/Request.java]: Null analysis was failing, so I had to initialize in the Request constructor.

**Comment by jkozlowski** [comment]
Need to loose the bogus APIs, revapi is getting weird: https://github.com/palantir/dialogue/pull/1107

File: dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachmentKey.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    private RequestAttachmentKey(Class<V> valueClazz) {
+        this.valueClazz = valueClazz;
+    }
+
+    void checkIsInstance(V value) {
+        assert valueClazz.isInstance(value) : "Value not instance of class " + valueClazz;
+    }
+
+    public static <T> RequestAttachmentKey<T> create(Class<T> valueClazz) {
```
Comment (carterkozak) [inline dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachmentKey.java]: Do we want `Class<? super T>` here for flexibility with generics?

Comment (jkozlowski) [reply dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachmentKey.java]: I couldn't get that to work, but maybe it was just late... Lemme try this again

Comment (jkozlowski) [reply dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachmentKey.java]: ```java: incompatible types: java.lang.Class<capture#1 of ? super T> cannot be converted to java.lang.Class<T>```

Comment (jkozlowski) [reply dialogue-target/src/main/java/com/palantir/dialogue/RequestAttachmentKey.java]: Alright, I think we got it.

File: dialogue-target/src/main/java/com/palantir/dialogue/Request.java [hunk 4]
```diff
@@ -82,6 +85,14 @@ private Request(Builder builder) {
         return body;
     }
 
+    /**
+     * The mutable request attachments for this request. Attachments will be propagated when this request is mutated
+     * through the builder
+     */
+    public RequestAttachments attachments() {
```
Comment (carterkozak) [inline dialogue-target/src/main/java/com/palantir/dialogue/Request.java:94]: May want to add a `@Beta` annotation here and on the `RequestAttachments` type for now.

**Review by carterkozak — APPROVED** [review]

File: dialogue-target/src/main/java/com/palantir/dialogue/Request.java (hunk 4 above)
Comment (jkozlowski) [reply dialogue-target/src/main/java/com/palantir/dialogue/Request.java:94]: Ack

**Pull request merged**